lib/gem_inclusions/storage.rb

```diff
@@ -0,0 +1,185 @@
+# frozen_string_literal: true
+
+class GemInclusions::Storage
+  HASH_PATTERN = /\A[0-9a-f]{64}\z/
+
+  FILE_KEY_FORMAT = "gems/%{gem}/contents/%{hash}"
+  SPEC_KEY_FORMAT = "gems/%{gem}/spec/%{full_name}.gemspec"
+  PATH_KEY_FORMAT = "gems/%{gem}/versions/%{version}/path/%{path}"
+  HASHES_KEY_FORMAT = "gems/%{gem}/hashes/%{version}"
```

Comment:
@simi The easiest way to explain how content is stored is to point at this snippet here.

Files are stored at the FILE_KEY at a hash of their content.
Specs are stored as the .gemspec file itself (this may not be quite complete, still adding tests)
Path keys store the actual file path and contain the hash, so you can get to a file contents with two S3 gets.
Hashes of all files in a gem version are stored for efficiency to allow yanking to not need to read every file in every gem on yank.